qupath-extension-openslide/src/main/java/qupath/lib/images/servers/openslide/OpenSlide.java

```diff
@@ -0,0 +1,491 @@
+
+    final public static String PROPERTY_NAME_BOUNDS_X = "openslide.bounds-x";
+
+    final public static String PROPERTY_NAME_BOUNDS_Y = "openslide.bounds-y";
+
+    final public static String PROPERTY_NAME_BOUNDS_WIDTH = "openslide.bounds-width";
+
+    final public static String PROPERTY_NAME_BOUNDS_HEIGHT = "openslide.bounds-height";
+
+    private long osr;
+
+    final private ReadWriteLock lock = new ReentrantReadWriteLock();
+
+    final private long[] levelWidths;
+
+    final private long[] levelHeights;
+
+    final private double[] levelDownsamples;
+
+    final private int levelCount;
+
+    final private Map<String, String> properties;
+
+    final private Map<String, AssociatedImage> associatedImages;
+
+    final private File canonicalFile;
+
+    final private int hashCodeVal;
+
+    public static String detectVendor(File file) {
+        return OpenSlideJNA.INSTANCE.openslide_detect_vendor(file.getPath());
+    }
+
+    public OpenSlide(File file) throws IOException {
+        if (!file.exists()) {
+            throw new FileNotFoundException(file.toString());
+        }
+
+        osr = OpenSlideJNA.INSTANCE.openslide_open(file.getPath());
+
+        if (osr == 0) {
+            throw new IOException(file
+                    + ": Not a file that OpenSlide can recognize");
+        }
+        // dispose on error, we are in the constructor
+        try {
+            checkError();
+        } catch (IOException e) {
+            dispose();
+            throw e;
+        }
+
+        // store level count
+        levelCount = OpenSlideJNA.INSTANCE.openslide_get_level_count(osr);
+
+        // store dimensions
+        levelWidths = new long[levelCount];
+        levelHeights = new long[levelCount];
+        levelDownsamples = new double[levelCount];
+
+        for (int i = 0; i < levelCount; i++) {
+            long[] w = new long[1], h = new long[1];
+            OpenSlideJNA.INSTANCE.openslide_get_level_dimensions(osr, i, w, h);
+            levelWidths[i] = w[0];
+            levelHeights[i] = h[0];
+            levelDownsamples[i] = OpenSlideJNA.INSTANCE.openslide_get_level_downsample(
+                    osr, i);
+        }
+
+        // properties
+        HashMap<String, String> props = new HashMap<>();
+        for (String s : OpenSlideJNA.INSTANCE.openslide_get_property_names(osr)) {
+            props.put(s, OpenSlideJNA.INSTANCE.openslide_get_property_value(osr, s));
+        }
+
+        properties = Collections.unmodifiableMap(props);
+
+        // associated images
+        HashMap<String, AssociatedImage> associated = new HashMap<>();
+        for (String s : OpenSlideJNA.INSTANCE
+                .openslide_get_associated_image_names(osr)) {
+            associated.put(s, new AssociatedImage(s, this));
+        }
+
+        associatedImages = Collections.unmodifiableMap(associated);
+
+        // store info for hash and equals
+        canonicalFile = file.getCanonicalFile();
```

Comment:
This is where having a file would help... but we can try creating one here if we really need it.